game-core/src/test/java/games/strategy/triplea/attachments/UnitTypeComparatorTest.java

```diff
@@ -0,0 +1,95 @@
+class UnitTypeComparatorTest {
+
+  private final UnitTypeComparator unitTypeComparator = new UnitTypeComparator();
+
+  @Mock private UnitType nullType;
+  @Mock private UnitType infrastructure;
+  @Mock private UnitType antiAircraft;
+  @Mock private UnitType air;
+  @Mock private UnitType sea;
+  @Mock private UnitType attacker;
+  @Mock private UnitType noneType1;
+  @Mock private UnitType noneType2;
+
+  @BeforeEach
+  void setUp() {
+    when(nullType.getName()).thenReturn("<NULL TYPE>");
+
+    final var infrastructureAttachment = mock(UnitAttachment.class);
+    when(infrastructureAttachment.getIsInfrastructure()).thenReturn(true);
+    when(infrastructure.getAttachment(any())).thenReturn(infrastructureAttachment);
+    when(infrastructure.getName()).thenReturn("<Infrastructure TYPE>");
+
+    final var antiAircraftAttachment = mock(UnitAttachment.class);
+    when(antiAircraftAttachment.getIsAaForBombingThisUnitOnly()).thenReturn(true);
+    when(antiAircraft.getAttachment(any())).thenReturn(antiAircraftAttachment);
+    when(antiAircraft.getName()).thenReturn("<AA TYPE>");
+
+    final var airAttachment = mock(UnitAttachment.class);
+    when(airAttachment.getIsAir()).thenReturn(true);
+    when(air.getAttachment(any())).thenReturn(airAttachment);
+    when(air.getName()).thenReturn("<Air TYPE>");
+
+    final var seaAttachment = mock(UnitAttachment.class);
+    when(seaAttachment.getIsSea()).thenReturn(true);
+    when(sea.getAttachment(any())).thenReturn(seaAttachment);
+    when(sea.getName()).thenReturn("<Sea Type>");
+
+    final var attackerAttachment = mock(UnitAttachment.class);
+    when(attackerAttachment.getAttack()).thenReturn(1);
+    when(attacker.getAttachment(any())).thenReturn(attackerAttachment);
+    when(attacker.getName()).thenReturn("<Attacker Type>");
+
+    when(noneType1.getAttachment(any())).thenReturn(mock(UnitAttachment.class));
+    when(noneType1.getName()).thenReturn("<NONE TYPE> - 1");
+
+    when(noneType2.getAttachment(any())).thenReturn(mock(UnitAttachment.class));
+    when(noneType2.getName()).thenReturn("<NONE TYPE> - 0");
+  }
+
+  @Test
+  void testUnitTypeComparator() {
+    assertThrows(
+        IllegalStateException.class, () -> unitTypeComparator.compare(nullType, noneType1));
+    assertThrows(
+        IllegalStateException.class, () -> unitTypeComparator.compare(noneType1, nullType));
+    assertThrows(IllegalStateException.class, () -> unitTypeComparator.compare(nullType, nullType));
+
+    assertEquals(0, unitTypeComparator.compare(infrastructure, infrastructure));
+    assertEquals(0, unitTypeComparator.compare(antiAircraft, antiAircraft));
+    assertEquals(0, unitTypeComparator.compare(air, air));
+    assertEquals(0, unitTypeComparator.compare(sea, sea));
+    assertEquals(0, unitTypeComparator.compare(attacker, attacker));
+    assertEquals(0, unitTypeComparator.compare(noneType1, noneType1));
+
+    assertEquals(1, unitTypeComparator.compare(infrastructure, antiAircraft));
```

Comment:
Noticing this is a problem in the prod code, why do we expect sea to be after attacker, or AA to be after air? I don't think it would be clear to a maintainer why we expect these conditions to be true, perhaps same issue in prod code and it's not clear why the ordering is defined as it is. Is it just arbitrary? Seems it would not be as the comparator defines a contract so that it is *not* arbitrary.

Thinking some assertion comments could help, eg:
```assertThat("AA should be after air because ...", unitTypeComparator.compare(antiAircraft, air), is(1))```;

Reply:
Well the javadocs of the UnitTypeComparator clearly state the order of the units.
Can't really tell you something about the reasoning